system/lib/wasmfs/paths.h

```diff
@@ -5,44 +5,71 @@
 
 #pragma once
 
-#include <vector>
+#include <cassert>
+#include <memory>
+#include <string_view>
+#include <variant>
 
 #include "file.h"
 
-namespace wasmfs {
+namespace wasmfs::path {
 
-struct ParsedPath {
-  std::optional<Directory::Handle> parent;
-  std::shared_ptr<File> child;
+// Typically -ENOTDIR or -ENOENT.
+using Error = long;
```

Comment:
Should this be `__wasi_errno_t`?

Reply:
Unclear! All the non-wasi syscalls use `long` as their return type.

Reply:
I think `__wasi_errno_t` is the way to go, unless I'm missing something. We already depend on some wasi error codes matching the values of musl ones. We can assume the type is int32_t as well. However, I am not sure why `long` is the type musl uses... that seems like it affects wasm64. Anyhow, not urgent to do in general, but maybe new code could at least use the wasi one?

Reply:
Ok, using wasi types and constants here sgtm. I'll do it as a NFC followup though and convert the entire code base at once.